core/src/main/java/org/apache/calcite/sql/SqlAsofJoin.java

```diff
@@ -0,0 +1,164 @@
+    this.matchCondition = matchCondition;
+  }
+
+  @SuppressWarnings("nullness")
+  @Override public List<SqlNode> getOperandList() {
+    return ImmutableNullableList.of(left, natural, joinType, right,
+        conditionType, condition, matchCondition);
+  }
+
+  @Override public SqlOperator getOperator() {
+    return ASOF_OPERATOR;
+  }
+
+  @SuppressWarnings("assignment.type.incompatible")
+  @Override public void setOperand(int i, @Nullable SqlNode operand) {
+    switch (i) {
+    case 0:
+      left = operand;
+      break;
+    case 1:
+      natural = (SqlLiteral) operand;
+      break;
+    case 2:
+      joinType = (SqlLiteral) operand;
+      break;
+    case 3:
+      right = operand;
+      break;
+    case 4:
+      conditionType = (SqlLiteral) operand;
+      break;
+    case 5:
+      condition = operand;
+      break;
+    case 6:
+      matchCondition = operand;
+      break;
+    default:
+      throw new AssertionError(i);
+    }
+  }
+
+  /**
+   * The match condition of the ASOF JOIN.
+   *
+   * @return The match condition of the ASOF join.
+   */
+  public final SqlNode getMatchCondition() {
+    return matchCondition;
+  }
+
+  /**
+   * Describes the syntax of the SQL ASOF JOIN operator.
+   */
+  public static class SqlAsofJoinOperator extends SqlOperator {
+    //~ Constructors -----------------------------------------------------------
+
+    private SqlAsofJoinOperator(String name, int prec) {
+      super(name, SqlKind.JOIN, prec, true, null, null, null);
+    }
+
+    //~ Methods ----------------------------------------------------------------
+
+    @Override public SqlSyntax getSyntax() {
+      return SqlSyntax.SPECIAL;
+    }
+
+    @SuppressWarnings("argument.type.incompatible")
+    @Override public SqlCall createCall(
+        @Nullable SqlLiteral functionQualifier,
+        SqlParserPos pos,
+        @Nullable SqlNode... operands) {
+      assert functionQualifier == null;
+      return new SqlAsofJoin(pos, operands[0], (SqlLiteral) operands[1],
+          (SqlLiteral) operands[2], operands[3], (SqlLiteral) operands[4],
+          operands[5], operands[6]);
+    }
+
+    @Override public void unparse(
+        SqlWriter writer,
+        SqlCall call,
+        int leftPrec,
+        int rightPrec) {
+      final SqlAsofJoin join = (SqlAsofJoin) call;
+
+      join.left.unparse(
+          writer,
+          leftPrec,
+          getLeftPrec());
+      if (join.getJoinType() == JoinType.LEFT_ASOF) {
+        writer.sep("LEFT ASOF JOIN");
+      } else {
+        writer.sep("ASOF JOIN");
+      }
+      join.right.unparse(writer, getRightPrec(), rightPrec);
+      SqlNode matchCondition = join.matchCondition;
+      writer.keyword("MATCH_CONDITION");
+      matchCondition.unparse(writer, 0, 0);
+
+      SqlNode joinCondition = join.condition;
+      if (joinCondition != null) {
+        switch (join.getConditionType()) {
+        case ON:
+          writer.keyword("ON");
```

Comment:
Does this restriction need to be processed in sql `Parser.jj`?

Reply:
I think it is, right now only ON is supported in the parser.
I plan to add USING later.